core/drivers/regulator/regulator.c

```diff
@@ -323,9 +303,116 @@ static TEE_Result regulator_core_cleanup(void)
 		}
 	}
 
-	regulator_print_state(__func__);
+	if (TRACE_LEVEL >= TRACE_DEBUG)
+		regulator_print_tree();
 
 	return TEE_SUCCESS;
 }
 
 release_init_resource(regulator_core_cleanup);
+
+/* Return updated message buffer position of NULL on failure */
+static __printf(3, 4) char *add_msg(char *cur, char *end, const char *fmt, ...)
+{
+	va_list ap = { };
+	int max_len = end - cur;
+	int ret = 0;
+
+	va_start(ap, fmt);
+	ret = vsnprintf(cur, max_len, fmt, ap);
+	va_end(ap);
+
+	if (ret < 0 || ret >= max_len)
+		return NULL;
+
+	return cur + ret;
+}
+
+static void __maybe_unused print_regulator(struct regulator *regulator,
+					   int indent)
+{
+	static const char * const level_unit[] = { "uV", "mV", "V" };
+	int max_unit = ARRAY_SIZE(level_unit);
+	int level_max = 0;
+	int level_min = 0;
+	int level_cur = 0;
+	char msg_buf[128] = { };
+	char *msg_end = msg_buf + sizeof(msg_buf);
+	char *msg = msg_buf;
+	int n_max = 0;
+	int n_min = 0;
+	int n_cur = 0;
+	int n = 0;
+
+	if (indent) {
+		for (n = 0; n < indent - 1; n++) {
+			msg = add_msg(msg, msg_end, "|   ");
+			if (!msg)
+				goto out;
+		}
+
+		msg = add_msg(msg, msg_end, "+-- ");
+		if (!msg)
+			goto out;
+	}
+
+	regulator_get_range(regulator, &level_min, &level_max);
+	level_cur = regulator_get_voltage(regulator);
+
+	for (n_cur = 1; !(level_cur % 1000) && n_cur < max_unit; n_cur++)
+		level_cur /= 1000;
+	for (n_max = 1; !(level_max % 1000) && n_max < max_unit; n_max++)
+		level_max /= 1000;
+	for (n_min = 1; !(level_min % 1000) && n_min < max_unit; n_min++)
+		level_min /= 1000;
+
+	msg = add_msg(msg, msg_end, "%s \t(%3s / refcnt %u / flags %#"PRIx32
+		      " / %d %s ", regulator_name(regulator),
+		      regulator_is_enabled(regulator) ? "on " : "off",
+		      regulator->refcount, regulator->flags,
+		      level_cur, level_unit[n_cur - 1]);
+	if (!msg)
+		goto out;
+
+	if (level_min == level_max)
+		msg = add_msg(msg, msg_end, "fixed)");
+	else if (level_max == INT_MAX)
+		msg = add_msg(msg, msg_end, "[%d %s .. MAX])",
+			      level_min, level_unit[n_min - 1]);
+	else
+		msg = add_msg(msg, msg_end, "[%d %s .. %d %s])",
+			      level_min, level_unit[n_min - 1],
+			      level_max, level_unit[n_max - 1]);
+
+out:
+	if (!msg)
+		snprintf(msg_end - 4, 4, "...");
+
+	IMSG("%s", msg_buf);
+}
+
+static void print_regulator_subtree(struct regulator *root __maybe_unused,
+				    int indent __maybe_unused)
+{
+#ifdef CFG_DRIVERS_REGULATOR_PRINT_TREE
+	struct regulator *regulator = NULL;
+
+	SLIST_FOREACH(regulator, &regulator_device_list, link) {
+		if (regulator->supply == root) {
+			print_regulator(regulator, indent + 1);
+			print_regulator_subtree(regulator, indent + 1);
```

Comment:
Would it be possible to do this without recursion?

Reply:
Yes, that's why recursion is discouraged. A comment explaining why this will not be a problem here might be sufficient. Asserting that `indent` isn't too large could also be an option.

Reply:
I'm ok with setting an upper limit to `indent`. Implementing this without recursion looks a bit painfull.
I'll append a fixup commit and add the same to `clk_print_tree()`.

Reply:
> Implementing this without recursion looks a bit painfull.

Oh, that sounds almost like a challenge. ;-)
You have parent pointers in the nodes so there's no need to save a list of pointers to find the path back to the root.

Reply:
It would be a bit of code to create the ascending tree of devices. Do you think it's worth it? This recursion is less than 10 lines, it like it. That said, I really understand your concern.

Reply:
I tried but I always end up with some sort of recursive calls to descend the tree to print the clock info.
The sole way out would be to allocate some big reference matrix to track per item / per level state of what has been printed so far. I found that quite ugly.

I would be in favor for the recursive calls, bounded to 16 levels (32 seems to high from the trace found on clock: https://github.com/OP-TEE/optee_os/pull/6489#issuecomment-1824730803).

Reply:
This should do the trick:
``` C
static struct regulator *find_next_regulator(struct regulator *parent,
					     struct regulator *sibling)
{
	struct regulator *r = NULL;

	if (sibling)
		r = SLIST_NEXT(sibling, link);
	else
		r = SLIST_FIRST(&regulator_device_list);

	while (r && r->supply != parent)
		r = SLIST_NEXT(r, link);

	return r;
}

static void regulator_print_tree(void)
{
	struct regulator *regulator = NULL;
	struct regulator *parent = NULL;
	struct regulator *next = NULL;
	int indent = -1;

	while (true) {
		next = find_next_regulator(parent, regulator);
		if (next) {
			print_regulator(next, indent + 1);
			/* Enter the subtree of the next regulator */
			parent = next;
			indent++;
			regulator = NULL;
		} else {
			/*
			 * We've processed all children at this level.
			 * If parent is NULL we're at the top and are done.
			 */
			if (!parent)
				break;
			/*
			 * Move up one level to resume with the next
			 * regulator of the parent.
			 */
			regulator = parent;
			parent = parent->supply;
			indent--;
		}
	}
}
```